README.md
# Sidekiq-Smart-Killer

[![Gem Version](https://badge.fury.io/rb/sidekiq-smart-killer.svg)](http://badge.fury.io/rb/sidekiq-smart-killer)
[![Build Status](https://github.com/ondrejbartas/sidekiq-smart-killer/workflows/CI/badge.svg?branch=master)](https://github.com/ondrejbartas/sidekiq-smart-killer/actions)
[![Coverage Status](https://coveralls.io/repos/github/ondrejbartas/sidekiq-smart-killer/badge.svg?branch=master)](https://coveralls.io/github/ondrejbartas/sidekiq-smart-killer?branch=master)

## Changelog

Before upgrading to a new version, please read our [Changelog](Changes.md).

## Installation

### Requirements

- Redis 2.8 or greater is required. (Redis 3.0.3 or greater is recommended for large scale use)
- Sidekiq 5, or 4, or 3 and greater is required (for Sidekiq < 4 use version sidekiq-smart-killer 0.3.1)

Install the gem:

```
$ gem install sidekiq-smart-killer
```

Or add to your `Gemfile` and run `bundle install`:

```ruby
gem "sidekiq-smart-killer"
```

**NOTE** If you are not using Rails, you need to add `require 'sidekiq-smart-killer'` somewhere after `require 'sidekiq'`.

## Getting Started

## Contributing

**Thanks to all [contributors](https://github.com/ondrejbartas/sidekiq-smart-killer/graphs/contributors), you’re awesome and this wouldn’t be possible without you!**

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so we don't break it in a future version unintentionally.

### Testing

You can execute the test suite by running:

```
$ bundle exec rake test
```

## License

Copyright (c) 2013 Ondrej Bartas. See [LICENSE](LICENSE.txt) for further details.


Changes.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.0.1

- Init


lib/sidekiq/smart_killer/history.rb
module Sidekiq
  module SmartKiller
    class History

      MINUTES_STEP = 10

      COUNT = 24 * 60 / MINUTES_STEP
      def self.hostnames
        Sidekiq.redis do |conn|
          conn.smembers("smart_killer:hostnames").compact.sort
        end
      end

      def self.mem_history(hostname, count = COUNT)
        history('mem', hostname, count)
      end

      def self.cpu_history(hostname, count = COUNT)
        history('cpu', hostname, count)
      end

      def self.processes(hostname)
        Sidekiq.redis do |conn|
          conn.hgetall("smart_killer:#{hostname}:processes").each_with_object({}) do |(identity, info), out|
            out[identity] = Sidekiq.load_json(info)
          end
        end
      end

      def self.history(kind, hostname, count = COUNT)
        start_date = Time.now.utc
        keys = []
        dates = []
        stat_hash = {}
        i = 0

        while i < count
          date = start_date - (i * MINUTES_STEP) * 60
          datestr = date.strftime("%Y-%m-%dT%H:%M:00")

          kk = []
          MINUTES_STEP.times do |y|
            d = date + 60 * y
            kk << "smart_killer:#{hostname}:#{kind}_usage:#{d.strftime("%Y-%m-%dT%H:%M:00")}"
          end

          stat_hash[datestr] = Sidekiq.redis do |conn|
            values = conn.mget(kk).compact.map(&:to_f)
            if values.any?
              values.sum / values.size
            else
              -1
            end
          end
          i += 1
        end

        stat_hash
      end
    end
  end
end


lib/sidekiq/smart_killer/launcher.rb
# require cron poller
require 'sidekiq/smart_killer/poller'

# For SmartKiller we need to add some methods to Launcher
# so look at the code bellow.
#
# we are creating new smart killer poller instance and
# adding start and stop commands to launcher
module Sidekiq
  module SmartKiller
    module Launcher
      # Add cron poller to launcher
      attr_reader :smart_killer_poller

      # add cron poller and execute normal initialize of Sidekiq launcher
      def initialize(options)
        @smart_killer_poller = Sidekiq::SmartKiller::Poller.new
        super(options)
      end

      # execute normal run of launcher and run cron poller
      def run
        super
        smart_killer_poller.start
      end

      # execute normal quiet of launcher and quiet cron poller
      def quiet
        smart_killer_poller.terminate
        super
      end

      # execute normal stop of launcher and stop cron poller
      def stop
        smart_killer_poller.terminate
        super
      end
    end
  end
end

Sidekiq.configure_server do
  # require  Sidekiq original launcher
  require 'sidekiq/launcher'
  puts "START"
  ::Sidekiq::Launcher.prepend(Sidekiq::SmartKiller::Launcher)
end


lib/sidekiq/smart_killer/web_extension.rb
module Sidekiq
  module SmartKiller
    module WebExtension

      def self.registered(app)

        app.settings.locales << File.join(File.expand_path("..", __FILE__), "locales")

        #index page of cron jobs
        app.get '/smart-killer' do
          view_path    = File.join(File.expand_path("..", __FILE__), "views")

          @cron_jobs = Sidekiq::Cron::Job.all

          #if Slim renderer exists and sidekiq has layout.slim in views
          if defined?(Slim) && File.exists?(File.join(settings.views,"layout.slim"))
            render(:slim, File.read(File.join(view_path, "smart_killer.slim")))
          else
            render(:erb, File.read(File.join(view_path, "smart_killer.erb")))
          end
        end

        app.post "/smart-killer/action" do
          if params['identity']
            p = Sidekiq::Process.new('identity' => params['identity'])
            p.quiet! if params['quiet']
            p.stop! if params['stop']
          else
            processes.each do |pro|
              pro.quiet! if params['quiet']
              pro.stop! if params['stop']
            end
          end

          redirect "#{root_path}smart-killer"
        end


      end
    end
  end
end


lib/sidekiq/smart_killer/poller.rb
require 'sidekiq'
require 'sidekiq/cron'
require 'sidekiq/scheduled'

module Sidekiq
  module SmartKiller
    POLL_INTERVAL = 30
    STOP_JOB_WHEN_MEMORY_OVER = 900
    MINIMUM_RUNNING_TIME = 30

    # The Poller checks host stats
    class Poller < Sidekiq::Scheduled::Poller
      STATS_TTL = 7*24*60*60

      def enqueue
        time = Time.now.utc

        Sidekiq.redis do |conn|
          if conn.set("smart_killer:getting_stats:#{hostname}", "started", ex: poll_interval_average - 10, nx: true)
            # puts "Started process"
          else
            # puts "Already working"
            return
          end
        end

        number_of_cpus = (`uname`.strip == 'Darwin' ? `sysctl -n hw.logicalcpu_max`.strip : `nproc`.strip).to_i
        x = `ps aux`.split("\n").map{|l| l.split(/\s+/) }
        y = x[1..-1].map{|l| Hash[x[0].zip(l)]}

        total_mem_usage = y.sum{|a| a['%MEM'].to_f }
        max_cpu_usage_processes =  y.sum{|a| a['%CPU'].to_f }
        total_cpu_usage = (max_cpu_usage_processes / (number_of_cpus*100.0))*100

        stats_by_pid = y.each_with_object({}) do |a, out|
          out[a['PID']] = a
        end

        host_processes = {}
        Sidekiq::ProcessSet.new.each do |process|
          next unless process['hostname'] == hostname
          pid = process['pid']
          host_processes[pid] = {
            process: process,
            stats: (stats_by_pid[pid.to_s] || {}),
            memory_in_mb: (stats_by_pid[pid.to_s] || {})['RSS'].to_i / 1024
          }
        end

        host_processes.
          select { |pid, v|
            v[:process]['busy'] == 0 &&
              v[:memory_in_mb] > STOP_JOB_WHEN_MEMORY_OVER &&
              v[:process]['started_at'] < MINIMUM_RUNNING_TIME.minutes.ago.to_i &&
              empty_queues?(v[:process]['queues'])
          }.each do |pid, v|

          puts "Stopping job #{v[:process].identity} with used memory #{v[:memory_in_mb]}MB"
          Sidekiq::Process.new('identity' => v[:process].identity).quiet!
          Sidekiq::Process.new('identity' => v[:process].identity).stop!
        end

        nowdate = Time.now.utc.strftime("%Y-%m-%dT%H:%M:00")
        Sidekiq.redis do |conn|
          conn.multi do
            conn.sadd("smart_killer:hostnames", hostname)
            mem_key = "smart_killer:#{hostname}:mem_usage:#{nowdate}"
            conn.set("#{mem_key}", total_mem_usage)
            conn.expire("#{mem_key}", STATS_TTL)

            cpu_key = "smart_killer:#{hostname}:cpu_usage:#{nowdate}"
            conn.set("#{cpu_key}", total_cpu_usage)
            conn.expire("#{cpu_key}", STATS_TTL)

            conn.del("smart_killer:#{hostname}:processes")
            host_processes.each do |pid, process|
              conn.hset(
                "smart_killer:#{hostname}:processes",
                process[:process].identity,
                serialize_process(process)
              )
            end
          end
        end

        {
          hostname: hostname,
          number_of_cpus: number_of_cpus,
          total_mem_usage: total_mem_usage,
          total_cpu_usage: total_cpu_usage,
          host_processes: host_processes,
        }
      rescue => ex
        # Most likely a problem with redis networking.
        # Punt and try again at the next interval
        Sidekiq.logger.error ex.message
        Sidekiq.logger.error ex.backtrace.first
        handle_exception(ex) if respond_to?(:handle_exception)
      end

      private

      def serialize_process(process)
        {
          memory_in_mb: process[:memory_in_mb],
          memory_percent: process[:stats]['%MEM'].to_f,
          cpu_percent: process[:stats]['%CPU'].to_f
        }.to_json
      rescue StandardError => e
        {
          memory_in_mb: -1,
          memory_percent: -1,
          cpu_percent: -1,
        }.to_json
      end

      def empty_queues?(queues)
        queues.all? { |queue| Sidekiq::Queue.new(queue).size == 0 }
      end

      def hostname
        @hostname ||= `hostname`.strip
      end

      def poll_interval_average
        30
      end
    end
  end
end
